File: prisma/tag-menu-item.ts
/**
 * Cuisine-templated dietary / allergen tags for demo menus.
 * Places ingest has no real allergen data — this is an approximation only.
 */
import {
  type AllergenId,
  type DietId,
  serializeTags,
} from "../src/lib/dietary";

export type TaggedMenuFields = {
  dietaryTags: string;
  allergens: string;
};

export type TagableMenuItem = {
  name: string;
  description: string;
  priceCents: number;
  image?: string;
  dietaryTags?: DietId[];
  allergens?: AllergenId[];
};

export type TagableCategory = {
  name: string;
  items: TagableMenuItem[];
};

export type TaggedMenuItem = Omit<TagableMenuItem, "dietaryTags" | "allergens"> &
  TaggedMenuFields;

export type TaggedCategory = {
  name: string;
  items: TaggedMenuItem[];
};

type TagInput = {
  name: string;
  description: string;
  dietaryTags?: DietId[];
  allergens?: AllergenId[];
  categoryName?: string;
  cuisineKey?: string;
};

const PEANUT_RE =
  /\b(peanut|peanuts|satay|massaman|crushed peanuts?|peanut sauce)\b/i;
const TREE_NUT_RE =
  /\b(almond|almonds|cashew|cashews|walnut|walnuts|pistachio|pistachios|hazelnut|hazelnuts|macadamia|pecan|pecans|tree.?nuts?)\b/i;
const VEGAN_RE = /\b(vegan|plant-based)\b/i;
const VEGETARIAN_RE =
  /\b(vegetarian|veggie|paneer|falafel|margherita|mushroom swiss|avocado toast|granola|edamame|seaweed salad|dal makhani|palak paneer|samosas?|onion bhaji|raita|guacamole|elote|garden salad|side salad|veggie bowl|veggie patty|risotto funghi|penne arrabbiata)\b/i;
const MEAT_RE =
  /\b(chicken|beef|pork|lamb|bacon|ham|salami|pepperoni|prawn|salmon|tuna|fish|squid|barramundi|flathead|meat|sausage|pancetta|katsu|gyoza|burrito|taco|quesadilla|parmigiana|bolognese|carbonara|satay|massaman|rogan josh|butter chicken)\b/i;
const GLUTEN_FREE_RE = /\b(gluten-?free|\bgf\b)\b/i;
const HALAL_RE = /\bhalal\b/i;
const DRINK_RE =
  /\b(ginger beer|lemonade|sparkling water|flat white|iced latte|fresh juice|hot chocolate|miso soup)\b/i;
const DAIRY_DRINK_RE =
  /\b(flat white|latte|cappuccino|macchiato|mocha|chai|hot chocolate|milkshake|thickshake|thai iced tea|milk|cream|yoghurt)\b/i;
const SAFE_SIDE_RE =
  /\b(jasmine rice|thick-cut chips|extra chips|chips|onion rings|garlic bread|edamame|seaweed salad|garden salad|side salad|raita|guacamole & chips)\b/i;

/** Derive demo dietary tags + allergens from item copy and optional overrides. */
export function tagMenuItem(input: TagInput): TaggedMenuFields {
  const haystack = `${input.name} ${input.description}`;
  const allergens = new Set<AllergenId>(input.allergens ?? []);
  const diets = new Set<DietId>(input.dietaryTags ?? []);

  if (PEANUT_RE.test(haystack)) allergens.add("peanuts");
  if (TREE_NUT_RE.test(haystack)) allergens.add("tree-nuts");

  if (VEGAN_RE.test(haystack)) {
    diets.add("vegan");
    diets.add("vegetarian");
  }
  if (VEGETARIAN_RE.test(haystack) && !MEAT_RE.test(haystack)) {
    diets.add("vegetarian");
  }
  if (GLUTEN_FREE_RE.test(haystack)) diets.add("gluten-free");
  if (HALAL_RE.test(haystack)) diets.add("halal");

  const cuisine = (input.cuisineKey ?? "").toLowerCase();
  if (
    cuisine === "indian" &&
    MEAT_RE.test(haystack) &&
    !/\bpork\b|\bbacon\b/i.test(haystack)
  ) {
    diets.add("halal");
  }

  const category = (input.categoryName ?? "").toLowerCase();
  const looksLikeDrink =
    category.includes("drink") ||
    category.includes("coffee") ||
    DRINK_RE.test(input.name);
  const looksLikeSafeSide =
    SAFE_SIDE_RE.test(input.name) ||
    (category.includes("side") &&
      !PEANUT_RE.test(haystack) &&
      !TREE_NUT_RE.test(haystack));

  // Conservative nut-free: only mark when clearly a drink/safe side or explicitly overridden,
  // and never when peanut/tree-nut allergens are present.
  if (allergens.has("peanuts") || allergens.has("tree-nuts")) {
    diets.delete("nut-free");
  } else if (
    diets.has("nut-free") ||
    looksLikeDrink ||
    looksLikeSafeSide ||
    /\bnut-?free\b/i.test(haystack)
  ) {
    diets.add("nut-free");
  }

  // Drinks in this demo catalog are gluten-free and halal, but milk-based ones
  // (flat white, latte, hot chocolate) are only vegetarian — never vegan.
  if (looksLikeDrink) {
    diets.add("vegetarian");
    diets.add("gluten-free");
    diets.add("halal");
    if (!DAIRY_DRINK_RE.test(haystack)) diets.add("vegan");
  }

  // Vegan implies vegetarian.
  if (diets.has("vegan")) diets.add("vegetarian");

  const orderedDiets: DietId[] = (
    ["vegan", "vegetarian", "gluten-free", "halal", "nut-free"] as const
  ).filter((id) => diets.has(id));
  const orderedAllergens: AllergenId[] = (
    ["peanuts", "tree-nuts"] as const
  ).filter((id) => allergens.has(id));

  return {
    dietaryTags: serializeTags(orderedDiets),
    allergens: serializeTags(orderedAllergens),
  };
}

/** Apply tagging across menu categories; returns venue-level dietary tag union. */
export function tagMenuCategories(
  categories: TagableCategory[],
  cuisineKey?: string,
): {
  categories: TaggedCategory[];
  restaurantDietaryTags: string;
} {
  const venueDiets = new Set<DietId>();
  const tagged: TaggedCategory[] = categories.map((cat) => ({
    name: cat.name,
    items: cat.items.map((item) => {
      const fields = tagMenuItem({
        name: item.name,
        description: item.description,
        dietaryTags: item.dietaryTags,
        allergens: item.allergens,
        categoryName: cat.name,
        cuisineKey,
      });
      for (const tag of JSON.parse(fields.dietaryTags) as DietId[]) {
        venueDiets.add(tag);
      }
      return {
        name: item.name,
        description: item.description,
        priceCents: item.priceCents,
        image: item.image,
        dietaryTags: fields.dietaryTags,
        allergens: fields.allergens,
      };
    }),
  }));

  const ordered = (
    ["vegan", "vegetarian", "gluten-free", "halal", "nut-free"] as const
  ).filter((id) => venueDiets.has(id));

  return {
    categories: tagged,
    restaurantDietaryTags: serializeTags(ordered),
  };
}

File: backend/app/domain/tag_menu_item.py
"""Cuisine-templated dietary / allergen tags for demo menus.

Ported from prisma/tag-menu-item.ts. Places ingest has no real allergen data —
this is an approximation only.
"""

from __future__ import annotations

import re
from typing import Any, Optional, Union

from app.domain.dietary import DIET_IDS, coerce_tag_list, tags_for_storage

PEANUT_RE = re.compile(
    r"\b(peanut|peanuts|satay|massaman|crushed peanuts?|peanut sauce)\b",
    re.I,
)
TREE_NUT_RE = re.compile(
    r"\b(almond|almonds|cashew|cashews|walnut|walnuts|pistachio|pistachios|"
    r"hazelnut|hazelnuts|macadamia|pecan|pecans|tree.?nuts?)\b",
    re.I,
)
VEGAN_RE = re.compile(r"\b(vegan|plant-based)\b", re.I)
VEGETARIAN_RE = re.compile(
    r"\b(vegetarian|veggie|paneer|falafel|margherita|mushroom swiss|avocado toast|"
    r"granola|edamame|seaweed salad|dal makhani|palak paneer|samosas?|onion bhaji|"
    r"raita|guacamole|elote|garden salad|side salad|veggie bowl|veggie patty|"
    r"risotto funghi|penne arrabbiata)\b",
    re.I,
)
MEAT_RE = re.compile(
    r"\b(chicken|beef|pork|lamb|bacon|ham|salami|pepperoni|prawn|salmon|tuna|fish|"
    r"squid|barramundi|flathead|meat|sausage|pancetta|katsu|gyoza|burrito|taco|"
    r"quesadilla|parmigiana|bolognese|carbonara|satay|massaman|rogan josh|"
    r"butter chicken)\b",
    re.I,
)
GLUTEN_FREE_RE = re.compile(r"\b(gluten-?free|\bgf\b)\b", re.I)
HALAL_RE = re.compile(r"\bhalal\b", re.I)
DRINK_RE = re.compile(
    r"\b(ginger beer|lemonade|sparkling water|flat white|iced latte|fresh juice|"
    r"hot chocolate|miso soup)\b",
    re.I,
)
DAIRY_DRINK_RE = re.compile(
    r"\b(flat white|latte|cappuccino|macchiato|mocha|chai|hot chocolate|"
    r"milkshake|thickshake|thai iced tea|milk|cream|yoghurt)\b",
    re.I,
)
SAFE_SIDE_RE = re.compile(
    r"\b(jasmine rice|thick-cut chips|extra chips|chips|onion rings|garlic bread|"
    r"edamame|seaweed salad|garden salad|side salad|raita|guacamole & chips)\b",
    re.I,
)
PORK_RE = re.compile(r"\bpork\b|\bbacon\b", re.I)
NUT_FREE_RE = re.compile(r"\bnut-?free\b", re.I)

TagInput = Optional[Union[list[str], str]]


def tag_menu_item(
    *,
    name: str,
    description: str,
    dietary_tags: TagInput = None,
    allergens: TagInput = None,
    category_name: Optional[str] = None,
    cuisine_key: Optional[str] = None,
) -> dict[str, list[str]]:
    """Derive demo dietary tags + allergens from item copy and optional overrides."""
    haystack = f"{name} {description}"
    allergen_set = set(coerce_tag_list(allergens))
    diet_set = set(coerce_tag_list(dietary_tags))

    if PEANUT_RE.search(haystack):
        allergen_set.add("peanuts")
    if TREE_NUT_RE.search(haystack):
        allergen_set.add("tree-nuts")

    if VEGAN_RE.search(haystack):
        diet_set.add("vegan")
        diet_set.add("vegetarian")
    if VEGETARIAN_RE.search(haystack) and not MEAT_RE.search(haystack):
        diet_set.add("vegetarian")
    if GLUTEN_FREE_RE.search(haystack):
        diet_set.add("gluten-free")
    if HALAL_RE.search(haystack):
        diet_set.add("halal")

    cuisine = (cuisine_key or "").lower()
    if cuisine == "indian" and MEAT_RE.search(haystack) and not PORK_RE.search(haystack):
        diet_set.add("halal")

    category = (category_name or "").lower()
    looks_like_drink = (
        "drink" in category or "coffee" in category or bool(DRINK_RE.search(name))
    )
    looks_like_safe_side = bool(SAFE_SIDE_RE.search(name)) or (
        "side" in category
        and not PEANUT_RE.search(haystack)
        and not TREE_NUT_RE.search(haystack)
    )

    # Conservative nut-free: only mark when clearly a drink/safe side or explicitly
    # overridden, and never when peanut/tree-nut allergens are present.
    if "peanuts" in allergen_set or "tree-nuts" in allergen_set:
        diet_set.discard("nut-free")
    elif (
        "nut-free" in diet_set
        or looks_like_drink
        or looks_like_safe_side
        or NUT_FREE_RE.search(haystack)
    ):
        diet_set.add("nut-free")

    # Drinks in this demo catalog are gluten-free and halal, but milk-based ones
    # are only vegetarian — never vegan.
    if looks_like_drink:
        diet_set.add("vegetarian")
        diet_set.add("gluten-free")
        diet_set.add("halal")
        if not DAIRY_DRINK_RE.search(haystack):
            diet_set.add("vegan")

    if "vegan" in diet_set:
        diet_set.add("vegetarian")

    ordered_diets = [tag for tag in DIET_IDS if tag in diet_set]
    ordered_allergens = [tag for tag in ("peanuts", "tree-nuts") if tag in allergen_set]
    return {
        "dietaryTags": tags_for_storage(ordered_diets),
        "allergens": tags_for_storage(ordered_allergens),
    }


def tag_menu_categories(
    categories: list[dict[str, Any]],
    cuisine_key: Optional[str] = None,
) -> tuple[list[dict[str, Any]], list[str]]:
    """Apply tagging across menu categories; returns venue-level dietary tag union."""
    venue_diets: set[str] = set()
    tagged: list[dict[str, Any]] = []

    for cat in categories:
        items_out: list[dict[str, Any]] = []
        for item in cat.get("items", []):
            fields = tag_menu_item(
                name=item["name"],
                description=item.get("description", ""),
                dietary_tags=item.get("dietaryTags"),
                allergens=item.get("allergens"),
                category_name=cat.get("name"),
                cuisine_key=cuisine_key,
            )
            for tag in fields["dietaryTags"]:
                venue_diets.add(tag)
            items_out.append(
                {
                    "name": item["name"],
                    "description": item.get("description", ""),
                    "priceCents": item["priceCents"],
                    "image": item.get("image"),
                    "dietaryTags": fields["dietaryTags"],
                    "allergens": fields["allergens"],
                }
            )
        tagged.append({"name": cat["name"], "items": items_out})

    ordered = [tag for tag in DIET_IDS if tag in venue_diets]
    return tagged, tags_for_storage(ordered)
